### PR TITLE
Publish Identity XDM Shared State (ECID, CustomIdentities)

### PR DESCRIPTION
**Pull request by nporter-adbe**
This PR adds support for sharing the `identityMap` as part of Identity's shared state. Any time that we update the standard shared state we also update the XDM shared state with a XDM representation of the current Identity data.

Sample Identity XDM shared state:
```json
{
  "identityMap" : {
    "ECID" : [
      {
        "id" : "59725191566662510105001351686447723202",
        "authenticationState" : "ambiguous"
      }
    ],
    "test-type" : [
      {
        "id" : "test-identifier",
        "authenticationState" : "authenticated",
        "primary" : false
      }
    ]
  }
}
```

### CODE REVIEW

File: AEPIdentity/Sources/IdentityMap.swift [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            return .loggedOut
+        default:
+            return .ambiguous
+        }
+    }
+}
+
+/// Defines a map containing a set of end user identities, keyed on either namespace integration code or the namespace ID of the identity.
+/// Within each namespace, the identity is unique. The values of the map are an array, meaning that more than one identity of each namespace may be carried.
+struct IdentityMap: Equatable {
```
Comment (nporter-adbe) [inline AEPIdentity/Sources/IdentityMap.swift:33]: Copied this over and made some minor changes such as renaming the auth state to `XDMAuthenticationState`.

File: AEPCore/Sources/eventhub/Extension.swift [hunk 2]
```diff
@@ -115,6 +115,39 @@ public extension Extension {
         return runtime.getSharedState(extensionName: extensionName, event: event, barrier: true)
     }
 
+    // MARK: - XDM Shared State
```
Comment (nporter-adbe) [inline AEPCore/Sources/eventhub/Extension.swift:118]: Forgot to add these in the previous XDM shared state PR.

**Comment by nporter-adbe** [comment]
@emdobrin @kevinlind need some feedback when time permits

File: AEPCore/Sources/eventhub/Extension.swift (hunk 2 above)
Comment (emdobrin) [reply AEPCore/Sources/eventhub/Extension.swift:118]: Can you split this in a separate PR so we can update dev with this and merge the Identity changes in a feature branch?

File: AEPIdentity/Sources/IdentityMap.swift [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    /// Adds an `IdentityItem` to this map. If an item is added which shares the same `namespace` and `id` as an item
+    /// already in the map, then the new item replaces the existing item.
+    /// - Parameters:
+    ///   - namespace: The namespace for this identity
+    ///   - id: Identity of the consumer in the related namespace.
+    ///   - authenticationState: The state this identity is authenticated as for this observed ExperienceEvent.
+    ///   - primary: Indicates if this identity is the preferred identity. It is used as a hint to help systems better organize how identities are queried.
+    mutating func addItem(namespace: String,
+                          id: String,
+                          authenticationState: MobileVisitorAuthenticationState? = nil,
```
Comment (emdobrin) [inline AEPIdentity/Sources/IdentityMap.swift:45]: I suggest to keep these concepts separated, and use XDMAuthenticationState here, this way we can use this API for pure xdm objects and maybe have some helper functions for conversions.

File: AEPIdentity/Sources/IdentityMap.swift [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                          primary: Bool? = nil) {
+        let item = IdentityItem(id: id,
```
Comment (emdobrin) [inline AEPIdentity/Sources/IdentityMap.swift:47]: We should check that namespace and id are non-empty strings.

File: AEPIdentity/Sources/IdentityProperties.swift [hunk 5]
```diff
@@ -63,6 +63,24 @@ struct IdentityProperties: Codable {
...
+        var identityMap = IdentityMap()
+
+        // Add ECID
+        if let ecidString = ecid?.ecidString {
+            identityMap.addItem(namespace: IdentityConstants.XDM.Keys.ECID, id: ecidString)
+        }
+
+        for customerId in customerIds ?? [] {
+            guard let type = customerId.type, let id = customerId.identifier else { continue }
+            identityMap.addItem(namespace: type, id: id, authenticationState: customerId.authenticationState, primary: false)
```
Comment (emdobrin) [inline AEPIdentity/Sources/IdentityProperties.swift:78]: This looks good for now for customer visitor ids. We are still confirming with the services team that this is supported for all customer id name, but we can revise this later when we have more details.

We will need to handle IDFA differently, which is currently stored in the customerIds list.

File: AEPIdentity/Sources/IdentityState.swift [hunk 6]
```diff
@@ -145,14 +145,17 @@ class IdentityState {
     ///   - response: the response data if any
     ///   - eventDispatcher: a function which when invoked dispatches an `Event` to the `EventHub`
     ///   - createSharedState: a function which when invoked creates a shared state for the Identity extension
-    func handleHitResponse(hit: IdentityHit, response: Data?, eventDispatcher: (Event) -> Void, createSharedState: ([String: Any], Event?) -> Void) {
+    ///   - createXDMSharedState: a function which when invoked creates a XDM shared state for the Identity extension
+    func handleHitResponse(hit: IdentityHit, response: Data?, eventDispatcher: (Event) -> Void, createSharedState: ([String: Any], Event?) -> Void,
+                           createXDMSharedState: ([String: Any], Event?) -> Void) {
```
Comment (emdobrin) [inline AEPIdentity/Sources/IdentityState.swift:150]: This would not be required if this path is never hit after the migration.

Comment (nporter-adbe) [reply AEPIdentity/Sources/IdentityState.swift:150]: Good catch, makes sense given my other PR now.

File: AEPCore/Sources/eventhub/Extension.swift (hunk 2 above)
Comment (nporter-adbe) [reply AEPCore/Sources/eventhub/Extension.swift:118]: Created https://github.com/adobe/aepsdk-core-ios/pull/450 and changed the target for this PR to the xdm feature branch. I'll get the XDM feature branch sync'd with dev once the change to dev land.

File: AEPIdentity/Tests/IdentityMapTests.swift [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        XCTAssertEqual(1, spaceItems?.count)
+        XCTAssertEqual("id", spaceItems?[0].id)
+        XCTAssertEqual("ambiguous", spaceItems?[0].authenticationState?.rawValue)
+        XCTAssertFalse(spaceItems?[0].primary ?? true)
+
+        let unknown = identityMap.getItemsFor(namespace: "unknown")
+        XCTAssertNil(unknown)
+    }
+
+    func testAddItems() {
```
Comment (kevinlind) [inline AEPIdentity/Tests/IdentityMapTests.swift:38]: Add a few tests for empty namespace and id to help keep the test coverage up. 😄

Comment (nporter-adbe) [reply AEPIdentity/Tests/IdentityMapTests.swift:38]: Added :)

**Comment by nporter-adbe** [comment]
@emdobrin @kevinlind feedback addressed, let me know if we're good to merge :)

**Review by kevinlind — APPROVED** [review]
Looks good for now. As @emdobrin pointed out, the IdentityMap namespace for IDFA is wrong, but we can update that later when we know the correct value.

File: AEPIdentity/Tests/IdentityStateTests.swift [hunk 8]
```diff
@@ -868,6 +879,8 @@ class IdentityStateTests: XCTestCase {
         // test
         state.processPrivacyChange(event: event, createSharedState: { _, _ in
             sharedStateExpectation.fulfill()
+        }, createXDMSharedState: { _, _ in
+            xdmSharedStateExpectation.fulfill()
```
Comment (emdobrin) [inline AEPIdentity/Tests/IdentityStateTests.swift:883]: We should also have some e2e tests to verify the values stored in the xdm shared state are correct after syncIdentifiers call or setPrivacyStatus changes, I am ok if you want to add that in a separate pr.

**Review by emdobrin — COMMENTED** [review]
@nporter-adbe changes look great, just added a suggestion for testing.

File: AEPIdentity/Tests/IdentityStateTests.swift (hunk 8 above)
Comment (nporter-adbe) [reply AEPIdentity/Tests/IdentityStateTests.swift:883]: Already merged :)
I'll get these additional tests in when we determine the final format of the namespace.
